config/vlan.py

```diff
@@ -117,103 +160,165 @@ def config_proxy_arp(db, vid, mode):
-            ctx.fail("{} is part of portchannel!".format(port))
-
-        if (clicommon.interface_is_untagged_member(db.cfgdb, port) and untagged): # TODO: MISSING CONSTRAINT IN YANG MODEL
-            ctx.fail("{} is already untagged member!".format(port))
+    if ADHOC_VALIDATION:
+        for vid in vid_list:
+
+            # default vlan checker
+            if vid == 1:
+                ctx.fail("{} is default VLAN".format(vlan))
+
+            log.log_info("'vlan member add {} {}' executing...".format(vid, port))
+
+            if not clicommon.is_vlanid_in_range(vid):
+                ctx.fail("Invalid VLAN ID {} (2-4094)".format(vid))
+
+            vlan = 'Vlan{}'.format(vid)
+            if clicommon.check_if_vlanid_exist(db.cfgdb, vlan) == False:
+                log.log_info("{} does not exist".format(vlan))
+                ctx.fail("{} does not exist".format(vlan))
+
+            if clicommon.get_interface_naming_mode() == "alias":  # TODO: MISSING CONSTRAINT IN YANG MODEL
+                alias = port
+                iface_alias_converter = clicommon.InterfaceAliasConverter(db)
+                port = iface_alias_converter.alias_to_name(alias)
+                if port is None:
+                    ctx.fail("cannot find port name for alias {}".format(alias))
+
+            # TODO: MISSING CONSTRAINT IN YANG MODEL
+            if clicommon.is_port_mirror_dst_port(db.cfgdb, port):
+                ctx.fail("{} is configured as mirror destination port".format(port))
+
+            # TODO: MISSING CONSTRAINT IN YANG MODEL
+            if clicommon.is_port_vlan_member(db.cfgdb, port, vlan):
+                log.log_info("{} is already a member of {}".format(port, vlan))
+                ctx.fail("{} is already a member of {}".format(port, vlan))
+                
+
+            if clicommon.is_valid_port(db.cfgdb, port):
+                is_port = True
+            elif clicommon.is_valid_portchannel(db.cfgdb, port):
+                is_port = False
+            else:
+                ctx.fail("{} does not exist".format(port))
+
+            if (is_port and clicommon.is_port_router_interface(db.cfgdb, port)) or \
+                    (not is_port and clicommon.is_pc_router_interface(db.cfgdb, port)):  # TODO: MISSING CONSTRAINT IN YANG MODEL
+                ctx.fail("{} is a router interface!".format(port))
+
+            portchannel_member_table = db.cfgdb.get_table('PORTCHANNEL_MEMBER')
+
+            # TODO: MISSING CONSTRAINT IN YANG MODEL
+            if (is_port and clicommon.interface_is_in_portchannel(portchannel_member_table, port)):
+                ctx.fail("{} is part of portchannel!".format(port))
+
+            # TODO: MISSING CONSTRAINT IN YANG MODEL
+            if (clicommon.interface_is_untagged_member(db.cfgdb, port) and untagged):
+                ctx.fail("{} is already untagged member!".format(port))
+            
+            # checking mode status of port if its access, trunk or routed
+            if is_port:
```

Comment:
I dont think, we need 'mode' field in the PORT table.

Please check VLAN_MEMBER table to know whether the port is untagged to only one VLAN (access mode) or port is untagged member to one VLAN and tagged member to one or more VLAN, then the port mode is trunk. if the port is not present in VLAN_MEMBER table, the mode is routed.

Reply:
Although having "switchport mode" gives more restriction control but it also takes away the liberty of freely modifying the vlan membership as needed... In my opinion, the argument could go both ways... But in my opinion it is better not to have to have the switchport mode then having it... My reason is that you should never give anyone the privilege to modify the VLAN configuration if they don't know what they are doing at the first place...  but again, this is arguable...  and for most people that came from the traditional L2 switching background and used to the concept of "Switchport" configuration this gives them the "familiar" flavor and feel of a NOS that they are used to...